Week3 improvements

snkirov opened the pull request:
As discussed, I am opening the PR without the pinch animation, which will be added with a future commit. Enjoy!

cameron-lightricks commented on ObjectiveCTesting/Modules/SetGame/Views/SetCardView.mm:
```diff
@@ -0,0 +1,228 @@
+      [self drawShapeAtOffset:-offsetForTwoShapes];
+      return;
+    case Three:
+      [self drawShapeAtOffset:offsetForOneShape];
+      [self drawShapeAtOffset:offsetForThreeShapes];
+      [self drawShapeAtOffset:-offsetForThreeShapes];
```
Was a bit surprised to see the `offsetForOneShape` being used in the three cases step. My suggestion would be to extract the switch to a method which just returns an array of offsets like:
```
  switch (_numberOfShapes) {
    case One:
      return [0];
    case Two:
      return [self.bounds.size.height / 8, -self.bounds.size.height / 8];
    case Three:
      return [self.bounds.size.height / 4, 0, -self.bounds.size.height / 4];
    case numberOfShapesCount:
      LogDebug(@"NumberOfShapes shouldn't be set to numberOfShapesCount in SetCardView.getColor");
      return [];
  }
```

And then iterate through the returned array drawing the shapes

cameron-lightricks commented on ObjectiveCTesting/Modules/SetGame/Views/SetCardView.mm:
```diff
@@ -0,0 +1,228 @@
+
+// MARK: - Fill
+
+- (void)fillShapeForPath:(UIBezierPath *) path {
+
+  [self strokeShapeFromPath:path];
```
Should this be called `colourShapeForPath` as it does `stroke` and `fill`? Also unneeded empty line :)

snkirov replied:
Did you mean `colorShapeForPath` hehe 😃 . Otherwise yeah, I like the idea!

cameron-lightricks commented on ObjectiveCTesting/Utils/CardMatching/PlayingCardMatchingService.mm:
```diff
@@ -0,0 +1,34 @@
+
+  int score = 0;
+
+  for (PlayingCard *otherCard in otherCards) {
+    if(playingCard.rank == otherCard.rank) {
```
Do we need to check if `otherCard` also is actually `PlayingCard`?

cameron-lightricks replied:
Also, need a space after `if`

snkirov replied:
Yeah we do, fixed the space as well

cameron-lightricks commented on ObjectiveCTesting/Utils/CardMatching/SetCardMatchingService.mm:
```diff
@@ -0,0 +1,116 @@
+  return hasMatched;
+}
+
+/// As per the rules of the game Set, we want to check whether the shapes of the given cards form a compoment match.
+/// Returns true if the shapes form a match and false otherwise.
+/// The algorithm is as follows.
+/// 1. Create an array with all possible shapes.
+/// 2. Iterate through each card and remove the coresponding shape from the array.
+/// 3. A match is formed if either all shapes have been removed from the array or if only one shape has been removed.
+- (BOOL)evaluateShapesForCards:(NSArray *)cards {
+  NSMutableArray *shapes = [[SetCardUtil shapesArray] mutableCopy];
+  auto shapesOriginalCount = shapes.count;
+  for (SetCard *card in cards) {
+    if ([shapes containsObject:[NSNumber numberWithInteger:card.shape]]) {
+      [shapes removeObject:[NSNumber numberWithInteger:card.shape]];
+    }
+  }
+  auto allCardsHaveTheSameShape = shapesOriginalCount - 1 == shapes.count;
+  auto allCardsHaveDifferentShapes = shapes.count == 0;
+  return allCardsHaveTheSameShape || allCardsHaveDifferentShapes;
+}
```
Nice! Very easy to follow 👍

cameron-lightricks commented on ObjectiveCTesting/Utils/CardMatching/SetCardMatchingService.mm:
```diff
@@ -0,0 +1,116 @@
+- (BOOL)evaluateFillsForCards:(NSArray *)cards {
+  NSMutableArray *fills = [[SetCardUtil fillsArray] mutableCopy];
+  auto fillsOriginalCount = fills.count;
+  for (SetCard *card in cards) {
+    if ([fills containsObject:[NSNumber numberWithInteger:card.fill]]) {
+      [fills removeObject:[NSNumber numberWithInteger:card.fill]];
+    }
+  }
```
If you wanted to, you could also use `[NSSet minusSet]` here: https://developer.apple.com/documentation/foundation/nsmutableset/1416710-minusset?language=objc

snkirov replied:
Feels like a smarter solution, but I think I want to stick with the stupid option here as well, for the sake of simplicity and making it explicit.

cameron-lightricks commented on ObjectiveCTesting/Utils/CardMatching/SetCardMatchingService.mm:
```diff
@@ -0,0 +1,116 @@
+/// 2. Iterate through each card and remove the coresponding color from the array.
+/// 3. A match is formed if either all colors have been removed from the array or if only one color has been removed.
+- (BOOL)evaluateColorsForCards:(NSArray *)cards {
+  NSMutableArray *colors = [[SetCardUtil colorsArray] mutableCopy];
+  auto colorsOriginalCount = colors.count;
+  for (SetCard *card in cards) {
+    if ([colors containsObject:[NSNumber numberWithInteger:card.color]]) {
+      [colors removeObject:[NSNumber numberWithInteger:card.color]];
+    }
+  }
+  auto allCardsHaveTheSameColor = colorsOriginalCount - 1 == colors.count;
+  auto allCardsHaveDifferentColors = colors.count == 0;
+  return allCardsHaveTheSameColor || allCardsHaveDifferentColors;
+}
```
If you wanted to remove some redundancy you could make the `colorsArray` an array of `int`, and map the `cards` to an array with the `int` value of their color property, and then extract the negation and `allSame || allDifferent` check to a separate method. This would let you re-use the logic for the matching for each of the different attributes.

BTW I'm not actually sure if this would be cleaner or nicer, so just something to consider, you do not have to make this change if you do not want to

cameron-lightricks replied:
makes sense!

cameron-lightricks commented on ObjectiveCTesting/Utils/CardUtils/PlayingCardUtil.h:
```diff
@@ -0,0 +1,18 @@
+@interface PlayingCardUtil : NSObject
+
+/// Returns an array with all valid suits.
+@property(class, nonatomic, readonly)NSArray *validSuits;
+/// Returns an array with all valid ranks as strings.
+@property(class, nonatomic, readonly)NSArray *rankStrings;
```
You can user `NSArray<Type *> *` to specify the types of the values you will be returning in the array. This is good practice as it held the compiler warn about any potential type clashes that could happen.

snkirov replied:
Fixed!

snkirov replied:
Done

cameron-lightricks commented on ObjectiveCTesting/Utils/CardUtils/PlayingCardUtil.mm:
```diff
@@ -0,0 +1,29 @@
+@implementation PlayingCardUtil
+
+static NSArray *_validSuits = @[@"♥️",@"♦️",@"♠️",@"♣️"];
+static NSArray *_rankStrings = @[@"?",@"A",@"2",@"3",@"4",@"5",@"6",
```
How come `"?"` is a valid rank string?

snkirov replied:
It's not but it is something that the lecture guy used and is a bit weird, but I kinda like it. The benefit is that the ranks map to their indexes, which is nice, although a bit surprising. I would leave it in if it doesn't bother you too much, even though the more I think about it, the more it goes against "good practices".

cameron-lightricks commented on ObjectiveCTesting/Utils/CardUtils/SetCardUtil.h:
```diff
@@ -0,0 +1,51 @@
+/// Utility class with helper methods for `SetCard`
+@interface SetCardUtil : NSObject
+/// Returns an array with all valid shapes.
+@property(class, nonatomic, readonly)NSArray *shapesArray;
+/// Returns an array with all valid colors.
+@property(class, nonatomic, readonly)NSArray *colorsArray;
+/// Returns an array with all valid fills.
+@property(class, nonatomic, readonly)NSArray *fillsArray;
+/// Returns an array with all valid numbers of shapes.
+@property(class, nonatomic, readonly)NSArray *numberOfShapesArray;
```
Again, try adding the type of objects in the array here

snkirov replied:
Fixed!

cameron-lightricks commented on ObjectiveCTesting/Utils/DeckFactory/DeckFactory.h:
```diff
@@ -0,0 +1,19 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class Deck;
+/// Factory class for deck generation.
+@interface DeckFactory : NSObject
```
Line gap needed here between forward declares and doc comment

cameron-lightricks reviewed (commented):
Great work 🏆 🙌  I think this is abstracted really neatly and is very easy to read. A lot of the code feels very simple, and most of my comments are just about style stuff. Good stuff 🏅

snkirov commented on ObjectiveCTesting.xcodeproj/project.pbxproj:
```diff
@@ -69,21 +69,102 @@
+			);
+			path = Models;
+			sourceTree = "<group>";
+		};
```
Fixed in later commits

snkirov commented on ObjectiveCTesting/Utils/CardGraphics/CardGraphicsProtocol.h:
```diff
@@ -0,0 +1,12 @@
+
+- (CGContext)drawCard:(Card *)card inContext:(CGContext *)context;
+
+@end
```
Fixed in later commits

snkirov commented on ObjectiveCTesting/Utils/CardMatching/CardMatchingProtocol.h:
```diff
@@ -0,0 +1,14 @@
+
+@end
+
+NS_ASSUME_NONNULL_END
```
Added documentation in later commits, but now added the array constraint as well.

snkirov commented on ObjectiveCTesting/Utils/DeckUtil.h:
```diff
@@ -0,0 +1,14 @@
+
+@interface DeckUtil : NSObject
++ (Card *)drawRandomCardFromDeck:(Deck *)deck;
+@end
```
Fixed this throughout the whole project.

snkirov commented on ObjectiveCTesting/Game Implementation/CardMatchingGame.mm:
```diff
@@ -5,15 +5,18 @@
+@property (nonatomic, readwrite) NSMutableString *lastMoveDescription;
 @property (nonatomic, strong) NSMutableArray<Card *> *cards;
 @property (nonatomic) NSInteger matchCount;
+@property (nonatomic, strong) PlayingCardMatchingService *matchingService;
```
Fixed in later commits.

snkirov commented on ObjectiveCTesting/Utils/CardMatching/PlayingCardMatchingService.mm:
```diff
@@ -0,0 +1,33 @@
+- (int)matchCard:(Card *)card withOtherCards:(NSArray *)otherCards {
+  if (![card isKindOfClass:[PlayingCard class]]) {
+    return 0;
+  }
```
Well, I was resting on the assumption that they would be the same type, but I guess I should check for the otherCards as well.

snkirov replied:
Done

snkirov commented on ObjectiveCTesting/Utils/DeckUtil.mm:
```diff
@@ -0,0 +1,20 @@
+  if ([deck.cards count]) {
+    unsigned index = arc4random() % [deck.cards count];
+    randomCard = deck.cards[index];
+    [deck.cards removeObjectAtIndex: index];
```
fixed in later commits.

snkirov commented on ObjectiveCTesting/Base.lproj/Main.storyboard:
```diff
@@ -268,30 +268,18 @@
                                     <action selector="resetGame:" destination="BYZ-38-t0r" eventType="touchUpInside" id="RHb-Tm-I71"/>
                                 </connections>
                             </button>
-                            <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="" textAlignment="center" lineBreakMode="characterWrap" numberOfLines="0" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="nF9-3Y-BL9">
-                                <rect key="frame" x="39" y="732" width="350" height="20"/>
-                                <constraints>
-                                    <constraint firstAttribute="height" constant="20" id="2oS-GA-pkE"/>
-                                    <constraint firstAttribute="width" constant="350" id="Ql7-cl-XfC"/>
-                                </constraints>
-                                <fontDescription key="fontDescription" type="system" pointSize="17"/>
-                                <nil key="textColor"/>
-                                <nil key="highlightedColor"/>
-                            </label>
```
" Your application this week is still required to play both the Set and Playing Card
matching games (in separate tabs) and must show the score and allow re-deals, but you
can remove the UI for showing the result of the last card choice as well as the History
MVC added last week. "

For some reason we didn't. I found it weird as well

snkirov commented on ObjectiveCTesting.xcodeproj/project.pbxproj:
```diff
@@ -112,6 +134,8 @@
 			isa = PBXGroup;
 			children = (
+				4FCA80E828B9098D00D962C6 /* PlayingCardGameViewController.h */,
+				4FCA80E928B9098D00D962C6 /* PlayingCardGameViewController.mm */,
```
definitely.

snkirov commented on ObjectiveCTesting.xcodeproj/project.pbxproj:
```diff
@@ -354,7 +590,7 @@
 				PRODUCT_BUNDLE_IDENTIFIER = com.snkirov.ObjectiveCTesting;
 				PRODUCT_NAME = "$(TARGET_NAME)";
 				SWIFT_EMIT_LOC_STRINGS = YES;
-				TARGETED_DEVICE_FAMILY = "1,2";
+				TARGETED_DEVICE_FAMILY = 1;
```
Yeah, I didn't want to add App Icons for iPad as well. Apparently I was being a bit lazy 😅

snkirov commented on ObjectiveCTesting/Assets.xcassets/AccentColor.colorset/Contents.json:
```diff
@@ -2,6 +2,15 @@
   "colors" : [
     {
       "idiom" : "universal"
+    },
+    {
+      "appearances" : [
+        {
+          "appearance" : "luminosity",
+          "value" : "dark"
+        }
+      ],
+      "idiom" : "universal"
```
No, but xcode complained if there was no accent color. Now I removed the accent color from the build settings as well and it is no longer complaining.

snkirov commented on ObjectiveCTesting/Game Implementation/CardMatchingGame+Protected.h:
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2022 Lightricks. All rights reserved.
+// Created by Svilen Kirov.
+
+#ifndef ObjectiveCTesting____FILEBASENAMEASIDENTIFIER_______FILEEXTENSION___
+#define ObjectiveCTesting____FILEBASENAMEASIDENTIFIER_______FILEEXTENSION___
```
Sometimes I missclick the file type when creating it. I think the Xcode objc templates have this include guards, whereas the LT files don't. I guess I could've gone through each file to make it consistent. So it's not intentional

snkirov commented on ObjectiveCTesting/Game Implementation/CardMatchingGame+Protected.h:
```diff
@@ -0,0 +1,18 @@
+@interface CardMatchingGame ()
+
+/// The underlying array of cards currently in play.
+@property (nonatomic, strong)NSMutableArray<Card *> *cardsInPlay;
```
Didn't know this. Fixed it everywhere.

snkirov commented on ObjectiveCTesting/Game Implementation/CardMatchingGame.mm:
```diff
@@ -0,0 +1,111 @@
+
+@implementation CardMatchingGame
+
+- (instancetype)initUsingDeck:(Deck *)deck {
+  return [self initUsingDeck:deck withMatchCount:2];
+}
```
You are right. No good reason for it to be here.

snkirov commented on ObjectiveCTesting/Game Implementation/CardMatchingGame.mm:
```diff
@@ -0,0 +1,111 @@
+}
+
+- (void)chooseCard:(Card *)card {
+
```
fixed in later commits.

snkirov commented on ObjectiveCTesting/Game Implementation/CardMatchingGame.mm:
```diff
@@ -0,0 +1,111 @@
+- (void)evaluateMatch:(Card *)card withCards:(NSArray<Card *> *)chosenCards {
+  auto matchScore = [self.matchingService matchCard:card withOtherCards:chosenCards];
+  if (matchScore) {
+    auto pointsForThisRound = matchScore * 4;
```
Definitely!

snkirov commented on ObjectiveCTesting/Game Implementation/PlayingCardMatchingGame.mm:
```diff
@@ -0,0 +1,44 @@
+
+@synthesize matchingService = _matchingService;
+
+- (instancetype)initUsingDeck:(Deck *)deck
+                   withMatchCount:(NSUInteger) matchCount {
```
Fixed it!

snkirov commented on ObjectiveCTesting/Game Implementation/PlayingCardMatchingGame.mm:
```diff
@@ -0,0 +1,44 @@
+  for (Card *card in self.cardsInPlay) {
+    if (![card isKindOfClass:[PlayingCard class]]) {
+      LogDebug(@"Card in getCardForSuit is not of kind PlayingCard.");
+      return nil;
```
Definitely, nice catch!

snkirov commented on ObjectiveCTesting/Game Implementation/SetCardMatchingGame.mm:
```diff
@@ -0,0 +1,54 @@
+}
+
+- (instancetype)initUsingDeck:(Deck *)deck
+                   withMatchCount:(NSUInteger) matchCount {
```
Fixed

snkirov commented on ObjectiveCTesting/Game Implementation/SetCardMatchingGame.mm:
```diff
@@ -0,0 +1,54 @@
+  for (Card *card in self.cardsInPlay) {
+    if (![card isKindOfClass:[SetCard class]]) {
+      LogDebug(@"Card in getCardForShape is not of kind SetCard.");
+      return nil;
```
Fixed!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Models/Deck.m:
```diff
@@ -0,0 +1,45 @@
+
+@implementation Deck
+
+- (instancetype) init {
```
Fixed it in other places as well.

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Models/Deck.m:
```diff
@@ -0,0 +1,45 @@
+}
+
+- (Card *)drawRandomCard {
+
```
gone!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Models/Deck.m:
```diff
@@ -0,0 +1,45 @@
+
+- (Card *)drawRandomCard {
+
+  if (![self.cards count]) {
```
nice catch!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/ViewControllers/BaseGameViewController.mm:
```diff
@@ -0,0 +1,496 @@
+  self.cardGrid.minCellHeight = 30 * 16 / 9;
+}
+
+- (void)evaluateDefaultGrid {
+  // The size is evaluated here, since the cardCanvas needs to be initialised to use it's frame.
+  self.cardGrid.size = self.cardCanvas.frame.size;
+  self.cardGrid.minimumNumberOfCells = defaultCardCount;
+}
+
+- (void)setupPlayingSection {
+  [self setupCardCanvas];
+  [self evaluateDefaultGrid];
+  [self setupCards];
+  [self.view layoutIfNeeded];
+}
```
Cool, will keep this in mind!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/ViewControllers/BaseGameViewController.mm:
```diff
@@ -0,0 +1,496 @@
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BaseGameViewController() <UIDynamicAnimatorDelegate>
+@property (nonatomic)BOOL shouldFloatCards;
```
Done.

snkirov commented on ObjectiveCTesting/Modules/BaseGame/ViewControllers/BaseGameViewController.mm:
```diff
@@ -0,0 +1,496 @@
+    return;
+  }
+
+  if (card.isChosen) {
+    [UIView transitionWithView:cardView
+                      duration:0.3
+                       options:cardView.animationOptionForTap
+                    animations:^{
+      [cardView setSelected:TRUE];
+    } completion:nil];
+    return;
+  }
+
+  if (cardView.selected)
+  [UIView transitionWithView:cardView
+                    duration:0.3
+                     options:UIViewAnimationOptionTransitionCrossDissolve
+                  animations:^{
+    [cardView setSelected:FALSE];
```
I renamed it for consistency. The idea here is that we want to make the views consistent with the models. If a card model isChosen, we want to select the card view. Otherwise, the model is not chosen, so we want to update the cardViews accordingly. I am open for ideas on how to make it less confusing!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/ViewControllers/BaseGameViewController.mm:
```diff
@@ -0,0 +1,496 @@
+- (void)drawMoreCards {
+  auto cardsPerDraw = 3;
+  for (int i = 0; i < cardsPerDraw; i++) {
+    auto cardView = [self generateCardView];
```
Makes perfect sense.

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Views/CardView.mm:
```diff
@@ -0,0 +1,76 @@
+  [self drawCardInterior];
+}
+
+- (instancetype)init {
+  if (self = [super init]) {
+    self.opaque = NO;
+    self.contentMode = UIViewContentModeRedraw;
+    self.selected = FALSE;
+  }
+  return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  if (self = [super initWithFrame:frame]) {
+    self.opaque = NO;
+    self.contentMode = UIViewContentModeRedraw;
+    self.selected = FALSE;
+  }
+  return self;
+}
```
Right! Fixed it.

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Views/CardView.mm:
```diff
@@ -0,0 +1,76 @@
+  [self drawCardInterior];
+}
+
+- (instancetype)init {
+  if (self = [super init]) {
+    self.opaque = NO;
+    self.contentMode = UIViewContentModeRedraw;
+    self.selected = FALSE;
+  }
+  return self;
+}
```
Nice, fixed it!

snkirov commented on ObjectiveCTesting/Modules/PlayingCardGame/Views/PlayingCardView.h:
```diff
@@ -0,0 +1,23 @@
+/// The suit of the underlying card.
+@property (nonatomic, readonly) NSString *suit;
+
+/// Sets up the view with the provided `SetCard`.
```
Whoopsies

snkirov commented on ObjectiveCTesting/Modules/PlayingCardGame/Views/PlayingCardView.mm:
```diff
@@ -0,0 +1,222 @@
+}
+
+- (void)drawCardInterior {
+  if (self.selected) {
+    [self drawFaceUpCard];
+    return;
+  }
+  [self drawFaceDownCard];
```
Cool will keep this in mind!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Models/Deck.h:
```diff
@@ -0,0 +1,29 @@
+
+/// Returns a random `Card` from the deck.
+/// Returns nil if the deck is empty.
+- (Card *)drawRandomCard;
```
Done!

snkirov commented on ObjectiveCTesting/Modules/BaseGame/ViewControllers/BaseGameViewController.mm:
```diff
@@ -0,0 +1,496 @@
+  textView.translatesAutoresizingMaskIntoConstraints = NO;
+
+  UILayoutGuide *guide = self.view.safeAreaLayoutGuide;
+  auto bottomConstraint = [textView.bottomAnchor
+                           constraintEqualToAnchor:guide.bottomAnchor
+                           constant:-edgeOffset];
+  auto centerConstraint = [textView.centerXAnchor constraintEqualToAnchor:guide.centerXAnchor];
```
I added a greaterThanOrEqual constraint between the different views in the bottom bar with an offset of 10.

snkirov commented on ObjectiveCTesting/Modules/BaseGame/Models/Card.m:
```diff
@@ -0,0 +1,12 @@
+//  Created by Svilen Kirov on 16/08/2022.
+//
+
+#import "Card.h"
+
+@implementation Card
+
+@end
```
Done!

snkirov commented on ObjectiveCTesting/Modules/PlayingCardGame/Views/PlayingCardView.mm:
```diff
@@ -0,0 +1,222 @@
+
+- (void)drawFaceUpCardWithImage:(UIImage *)faceImage {
+  const CGFloat defaultFaceCardScaleFactor = 0.90;
+  CGRect imageRect = CGRectInset(self.bounds,
+                                 self.bounds.size.width * (1.0 - defaultFaceCardScaleFactor),
+                                 self.bounds.size.height * (1.0 - defaultFaceCardScaleFactor));
```
Done!

snkirov commented on ObjectiveCTesting/Modules/PlayingCardGame/Views/PlayingCardView.mm:
```diff
@@ -0,0 +1,222 @@
+  auto centre = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2);
+  NSAttributedString *attributedSuit = [self generateAttributedSuit];
+  CGSize pipSize = [attributedSuit size];
+  auto pipOrigin = CGPointMake(centre.x - pipSize.width /
+                               pipOriginConstant - horizontalOffset * self.bounds.size.width,
+                               centre.y - pipSize.height /
+                               pipOriginConstant - verticalOffset * self.bounds.size.height);
```
If only I could, I would 😆. I tried understanding it, and it works for some reason, but I am not completely sure why. I feel like leaving it as it is, with a promise of not writing magical math in my own code.

cameron-lightricks replied:
Ah, I see that this is copy-pasted from the lecture then, no worries :)

snkirov commented:
I addressed every comment and implemented most of your suggestions. I would ask you to go once more through the unresolved comments, and add anything if needed. Afterwards hopefully we will be done with the Stanford Course! 🎉
